design-proposals/klip-19-materialize-views.md

```diff
@@ -0,0 +1,223 @@
+
+## Motivation and background
+
+### Why might it be bad to allow inserts into tables?
+
+Let's consider two examples:
+1. A table in an upstream db is being replicated to a topic in Kafka, e.g. via Connect, and users want this 
+   table available in KSQL.
+1. A user wants to create a table in KSQL that is derived from other source(s), as currently done using a 
+   `CREATE TABLE AS SELECT` statement.
+   
+In the first example, the _source of truth_ for the data set is the upstream db, not in Kafka. Any change to the data 
+needs to be done in the upstream db. Simply writing a new row to the Kafka topic is semantically incorrect. The row will
+be overwritten if/when the same row in the database is next updated. What's more, that same topic may be used by other 
+systems that expect it to be an exact copy of the data in the database table. 
+
+In the second example, the rows written to the sink topic are a _derived dataset_.  Simply writing a new row to the Kafka 
+topic is also semantically incorrect. Any row written would potentially be overwritten by an upstream change.
+
+### When is it OK to insert into a table?
+
+We _can_ allow inserts into tables where KSQL is the only writer of that data and we can ensure correct table semantics.
+
+At the moment, users can create a new table in KSQL using a `CREATE TABLE` statement with `PARTITIONS` and, optionally,
+`REPLICAS`. This will create a new topic. As no other system writes to this topic, KSQL can support `INSERT`s into it.
+Likewise, a `CREATE TABLE` statement can be used to import an existing topic into KSQL. As long as no other system is 
+writing to that topic, (and that includes another KSQL topology), KSQL can offer correct `INSERT` semantics.
+
+The defining factor here is that it is only possible to have correct `INSERT` semantics _if no other system updates the table_.
+
+**Side note**: It would be easy to support correct `DELETE` semantics too. However, correct `UPDATE` row semantics will
+require more work. Likely, it will involve proxying calls to the KSQL node that owns the key, ensuring a single writer
+pattern.
+
+### How can we express this difference in SQL?   
+
+To ensure correct table semantics KSQL needs to differentiate between sources where it is the sole writer and those updated
+by other systems. Tables where KSQL is the only writer are _source datasets_. Topic containing data updated by another system, 
+including another KSQL query, should be treated as _read-only_ as KSQL can not offer correct table semantics on `INSERT`s.
+
+SQL already has concepts to differentiate: 
+ * a `TABLE` is a source dataset. The data in the table can be mutated via `INSERT`/`UPDATE`/`DELETE`s.
+ * a `VIEW` is a derived and read-only dataset. It's state is _derived_ from one or more sources. 
+  `INSERT`/`UPDATE`/`DELETE`s are not supported. 
+ * A `MATERIALIZED VIEW` is a `VIEW` where the cost of calculating the derived dataset is done on the write path, rather 
+   than the read.
```

Comment:
I think we need to clarify what `MATERIALIZED` implies, because in ksqlDB (in contrast to a RDBMS) it could mean to write into a topic (output or changelog?) and/or state store.

Given the remainder of the KLIP, I assume you propose that `MATERIALIZED` only implies that there will be an output topic, and it does _not_ imply anything with regard to state stores? I am not 100% sure atm, especially given "pull queries", if those semantics will be confusion to users (ie, "it's materialized, why can't I query it?" or later "it's materialized, why are my queries so slow"). To be fair, I don't have a better proposal atm, but want to raise this point as it seems important.

Reply:
Yes, `MATERIALIZED` is pertaining to the data being _materialized_ in a topic.

IMHO, whether a table is also materialized into a state store / rocksdb inside KSQL would ideally be an implementation detail.  The engine may in the future _always_ materialize tables into a rocks db store, or choose to materialize only heavily used tables, etc.  However, crucially, this should be hidden from normal users.  More importantly, I think this is outside the scope of this discussion - though I agree the KLIP should be more clear.

Reply:
To avoid confusion on this point, perhaps we should avoid using the term "materialized" internally in the code to mean "state stored in rocksdb" as the new materialized keyword will have a different meaning.